doc/rst/technotes/dsi.rst

```diff
@@ -954,5 +954,185 @@ implicitly for each ``GlobalDomain`` object and so will be local in any case.
+---------------------------------------------------
+
+The bulk-transfer interface is an optional interface supported only on array
+descriptors. The following methods are prefixed with ``doi`` for "Domain map
+Optional Interface." An array descriptor can implement one or more of these
+methods to opt-in to support for bulk-transfers:
+
+.. method:: proc GlobalArray.doiBulkTransferToKnown(myDom:domain, otherClass, otherDom:domain) : bool
+
+  Perform an array assignment from ``this`` to ``otherClass``.
+
+.. method:: proc GlobalArray.doiBulkTransferFromKnown(myDom:domain, otherClass, otherDom:domain) : bool
+
+  Perform an array assignment from ``otherClass`` to ``this``
+
+.. method:: proc GlobalArray.doiBulkTransferToAny(myDom:domain, otherClass, otherDom:domain) : bool
+
+  Perform an array assignment from ``this`` to ``otherClass``.
+
+.. method:: proc GlobalArray.doiBulkTransferFromAny(myDom:domain, otherClass, otherDom:domain) : bool
+
+  Perform an array assignment from ``otherClass`` to ``this``
+
+Each method shares these arguments:
+
+* ``myDom:domain`` - the indices to be transferred to/from ``this``
+
+* ``otherClass`` - a class inheriting from ``BaseArr`` that represents the
+  other array in the assignment
+
+* ``otherDom:domain`` - the indices to be transferred to/from ``otherClass``
+
+Callers of these methods must satisfy the following guarantees for these
+arguments:
+
+* ``this.rank == myDom.rank``
+
+* ``otherClass.rank == otherDom.rank``
+
+* ``myDom.size == otherDom.size``
+
+.. note::
+
+  ``myDom`` and ``otherDom`` are not required to be distributed, even if the
+  corresponding arrays are distributed. They simply represent the indices
+  involved in the transfer.
+
+Each method returns a ``bool`` indicating whether the assignment occurred. This
+allows domain maps to resort to the Chapel implementation of array assignment
+if they are incapable or unwilling to perform the assignment themselves.
+
+Known and Any methods
+---------------------
+
+There are two kinds of transfers accounted for in this interface.
+
+The first kind of transfer is a ``Known`` transfer. These are transfers that
+the implementer believes to be optimal. Typically this means that the
+implementer knows the type and internals of the other domain map involved in
+the transfer.
+
+The second kind of transfer is an ``Any`` transfer. These are transfers to/from
+domain maps whose type and internals are **not** known. These transfers rely on
+a common interface, which may involve unnecessary overhead compared to a
+``Known`` transfer. For example, it is likely that many domain maps will
+implement ``Known`` transfers to/from Chapel's default rectangular arrays. It
+is also likely that many domain maps will be implemented using default
+rectangular arrays. Knowing this, implementers of ``Any`` transfers can
+implement in terms of ``Known`` transfers to/from their own default rectangular
+arrays as a common interface.
```

Comment:
I'm having a little trouble grokking this without an example.  I'm thinking the idea is that I'm writing ``FancyDist``, and array ``A`` distributed using it will have its data stored in ``this.myArray[]``, a ``DefaultRectangular``.  Then ``FancyDistArrayDescriptor.transferFromAny(myDom, otherClass, otherDom)`` can figure it's likely ``otherClass`` implements ``Known`` transfers relative to ``DefaultRectangular``, so it can call ``otherClass.transferToKnown(otherDom, this.myArray, someTransform(myDom))`` (assuming it resolves) and if it returns true, be done?

Can ``FancyDist``'s storage of the elements in ``myArray[]`` require multiple calls to ``otherClass.transferToKnown()`` on different subsets of the domains?  Can it assume that if the first call returns true, that the subsequent calls to the same ``otherClass.transferToKnown()`` (passing the same array and having domains of comparable kind as the first call) will return true as well?  That is, is it bad / a non-issue if ``otherClass.transferToKnown()`` returns false for some cases that are type-wise the same, e.g. because of alignment or the provided domain has more elements than it's prepared to deal with, or so few that it doesn't think it's worthwhile to do "in bulk", or the transfer has some allocation or trylock step that could fail?

Reply:
Yes, I think you have the right idea when it comes to how one could implement a transfer in terms of DefaultRectangular. Personally I would like to add some helper functions to aid domain map authors in writing such transfers, but I haven't discussed that with anyone and don't really know what it would look like.

The rule for calling the transferTo/From methods is that the caller must complete the transfer if the returned value is false. So if a call for a given locale's data returns false, then the domain map author should write something like:
```chpl
for (i, j) in zip(myDom, otherDom) do
  localData[i] = otherClass.dsiAccess(j);
```

I don't want to add a full tutorial on writing this kind of transfer, but I recognize that documentation I've written isn't very helpful in this area. I'll see if I can add a brief section on ``Any`` transfers.